### PR TITLE
Add VerifyVersionedTransition replication task

### PR DESCRIPTION
**Pull request by xwduan**
## What changed?

Add VerifyVersionedTransition replication task
## Why?

For state replication project
## How did you test it?

unit test
## Potential risks

no risk. Not in use for now.
## Documentation

n/a
## Is hotfix candidate?

no

### CODE REVIEW

File: service/history/replication/executable_backfill_history_events_task.go [hunk 1]
```diff
@@ -203,43 +203,6 @@ func (e *ExecutableBackfillHistoryEventsTask) HandleErr(err error) error {
 	}
 }
 
-func (e *ExecutableBackfillHistoryEventsTask) MarkPoisonPill() error {
```
Comment (xwduan) [inline service/history/replication/executable_backfill_history_events_task.go:206]: This moved to ExecutableTask, so we don't need to copy logic into every task.

File: proto/internal/temporal/server/api/replication/v1/message.proto [hunk 2]
```diff
@@ -213,3 +214,11 @@ message NewRunInfo {
     string run_id = 1;
     temporal.api.common.v1.DataBlob event_batch = 2;
 }
+
+message VerifyVersionedTransitionTaskAttributes {
+    string namespace_id = 1;
+    string workflow_id = 2;
+    string run_id = 3;
+    int64 next_event_id = 4;
```
Comment (yycptt) [inline proto/internal/temporal/server/api/replication/v1/message.proto:222]: I think I was wrong. The event version might not equal to the version in the transition history. We probably should explicitly send the last event version history item.

File: service/history/replication/executable_verify_versioned_transition_task.go [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+			)
+		default:
+			return err
+		}
+	}
+	verifyNewRun := func() error {
+		if len(e.taskAttr.NewRunId) == 0 {
+			return nil
+		}
+		_, err = engine.GetMutableState(ctx, &historyservice.GetMutableStateRequest{
```
Comment (yycptt) [inline service/history/replication/executable_verify_versioned_transition_task.go:157]: is it possible to directly load mutable state here?

File: service/history/replication/executable_backfill_history_events_task.go (hunk 1 above)
Comment (yycptt) [reply service/history/replication/executable_backfill_history_events_task.go:206]: nice!

**Review by yycptt — COMMENTED** [review]
Haven't review the actual verify logic yet. 🙏

File: proto/internal/temporal/server/api/replication/v1/message.proto (hunk 2 above)
Comment (xwduan) [reply proto/internal/temporal/server/api/replication/v1/message.proto]: From offline discussion, we don't need the version here. But after some thought, I think we still need last event version history item to get event version to determine if the non-current branch has the events. The state transition on the verified task may not generate events, so we cannot get the last event version from the state transition failover version.

File: service/history/replication/executable_verify_versioned_transition_task.go [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+	}
+
+	_, err = versionhistory.FindFirstVersionHistoryIndexByVersionHistoryItem(ms.GetExecutionInfo().VersionHistories, e.taskAttr.LastVersionHistoryItem)
+	// case 2: events are to date on non-current branch
+	if err == nil {
+		return e.verifyNewRunExist(ctx)
+	}
+
+	// case 3: verify task has newer versioned transition, need to sync state
+	if workflow.CompareVersionedTransition(e.ReplicationTask().VersionedTransition, transitionHistory[len(transitionHistory)-1]) > 0 {
```
Comment (yycptt) [inline service/history/replication/executable_verify_versioned_transition_task.go:155]: I think we need to do this check before case 2.

File: proto/internal/temporal/server/api/replication/v1/message.proto (hunk 2 above)
Comment (yycptt) [reply proto/internal/temporal/server/api/replication/v1/message.proto]: not sure I follow. let's sync offline again.

File: service/history/replication/executable_task_tool_box.go [hunk 5]
```diff
@@ -60,5 +61,7 @@ type (
 		EventSerializer          serialization.Serializer
 		DLQWriter                DLQWriter
 		HistoryEventsHandler     eventhandler.HistoryEventsHandler
+		WorkflowCache            wcache.Cache
```
Comment (yycptt) [inline service/history/replication/executable_task_tool_box.go:64]: we need to make sure host level workflow cache is always enabled. 
I will check if we can deprecate shard level cache code path in next release.

**Review by yycptt — COMMENTED** [review]
LGTM overall. Some minor comments/questions

File: service/history/replication/raw_task_converter.go [hunk 6]
```diff
@@ -326,12 +324,38 @@ func convertSyncVersionedTransitionTask(
...
+			return &replicationspb.ReplicationTask{
+				TaskType:     enumsspb.REPLICATION_TASK_TYPE_VERIFY_VERSIONED_TRANSITION_TASK,
+				SourceTaskId: taskInfo.TaskID,
+				Attributes: &replicationspb.ReplicationTask_VerifyVersionedTransitionTaskAttributes{
+					VerifyVersionedTransitionTaskAttributes: &replicationspb.VerifyVersionedTransitionTaskAttributes{
+						NamespaceId:         taskInfo.NamespaceID,
+						WorkflowId:          taskInfo.WorkflowID,
+						RunId:               taskInfo.RunID,
+						NewRunId:            taskInfo.NewRunID,
+						EventVersionHistory: mutableState.GetExecutionInfo().VersionHistories.Histories[versionHistoryIndex].GetItems(),
```
[the remark this answers is not in the document]
Comment (xwduan) [reply service/history/replication/raw_task_converter.go]: Fixed to use pointer directly.

File: service/history/replication/executable_verify_versioned_transition_task.go [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+func (e *ExecutableVerifyVersionedTransitionTask) verifyNewRunExist(ctx context.Context) error {
+	if len(e.taskAttr.NewRunId) == 0 {
+		return nil
+	}
+	_, err := e.getMutableState(ctx, e.taskAttr.NewRunId)
+	switch err.(type) {
+	case nil:
+		return nil
+	case *serviceerror.NotFound:
+		return serviceerror.NewDataLoss(fmt.Sprintf("workflow new run not found. NamespaceId: %v, workflowId: %v, runId: %v, newRunId: %v",
```
[the remark this answers is not in the document]
Comment (xwduan) [reply service/history/replication/executable_verify_versioned_transition_task.go:204]: Yes, and if the original events are applied and new run was not found, then it means it is a data lose.

File: service/history/replication/executable_verify_versioned_transition_task.go [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+			"mutable state not up to date",
+			e.NamespaceID,
+			e.WorkflowID,
+			e.RunID,
+			e.ReplicationTask().VersionedTransition,
+		)
+	}
+	// case 3: state transition is not on non-current branch, but no event to verify
+	if e.taskAttr.NextEventId <= 1 {
+		return nil
```
[the remark this answers is not in the document]
Comment (xwduan) [reply service/history/replication/executable_verify_versioned_transition_task.go]: fixed.

**Review by yycptt — APPROVED** [review]

**Pull request merged**